File: frontend/src/api/methodImport.ts
import { apiClient } from './client';

export interface ParsedMethod {
  instrument: string | null;
  method_name: string | null;
  column_temp_c: number | null;
  flow_rate_ml_min: number | null;
  solvent_a: string | null;
  solvent_b: string | null;
  method_end_time_min: number | null;
  injection_volume_ul: number | null;
  sampler_temp_c: number | null;
  percent_b_start: number | null;
  percent_b_end: number | null;
  gradient_time_min: number | null;
  gradient_table: Array<{
    time_min: number;
    flow_rate_ml_min: number | null;
    percent_b: number | null;
    curve: number | null;
  }>;
  warnings: string[];
}

export interface MzXmlSummary {
  filename: string;
  num_scans: number;
  num_ms1_scans: number;
  num_ms2_scans: number;
  rt_start_s: number | null;
  rt_end_s: number | null;
  polarity: string | null;
}

export interface PeakDetectionResult {
  compound_id: string | null;
  compound_name: string | null;
  smiles: string | null;
  target_mz: number | null;
  mz_tolerance_ppm: number | null;
  peaks: Array<{
    retention_time_s: number;
    retention_time_min: number;
    intensity: number;
    peak_width_s: number | null;
    signal_to_noise: number | null;
  }>;
  xic_points: number;
  error: string | null;
}

export interface ExtractPeaksResponse {
  mzxml_summaries: MzXmlSummary[];
  results: PeakDetectionResult[];
  method_conditions: ParsedMethod | null;
}

export interface TrainFromPeaksResponse {
  artifact_id: string;
  n_samples: number;
  n_new_samples: number;
  column_type: string;
  model_type: string;
  compounds_used: string[];
  compounds_no_peaks: string[];
  incremental: boolean;
  existing_samples_loaded?: number;
  existing_model_version?: number;
}

export interface ModelSummary {
  id: string;
  column_type: string;
  model_type: string;
  version: number;
  n_samples: number;
  trained_at: string;
}

export interface MethodConditionOverrides {
  override_flow?: number;
  override_temp?: number;
  override_percent_b_start?: number;
  override_percent_b_end?: number;
  override_gradient_time?: number;
  override_ph?: number;
}

export const methodImportApi = {
  parseMeth: async (file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    const { data } = await apiClient.post<ParsedMethod>('/method-import/parse-meth', formData, {
      headers: { 'Content-Type': undefined },
    });
    return data;
  },

  parseMzxml: async (file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    const { data } = await apiClient.post<MzXmlSummary>('/method-import/parse-mzxml', formData, {
      headers: { 'Content-Type': undefined },
    });
    return data;
  },

  listModels: async (columnType?: string) => {
    const { data } = await apiClient.get<ModelSummary[]>('/method-import/models', {
      params: columnType ? { column_type: columnType } : undefined,
    });
    return data;
  },

  extractPeaks: async (
    mzxmlFiles: File[],
    compoundIds: string[],
    options?: {
      methFile?: File;
      mz_tolerance_ppm?: number;
      min_snr?: number;
      max_peaks_per_compound?: number;
    },
  ) => {
    const formData = new FormData();
    for (const f of mzxmlFiles) {
      formData.append('mzxml_files', f);
    }
    formData.append('compound_ids', compoundIds.join(','));
    if (options?.methFile) {
      formData.append('meth_file', options.methFile);
    }
    if (options?.mz_tolerance_ppm) {
      formData.append('mz_tolerance_ppm', String(options.mz_tolerance_ppm));
    }
    if (options?.min_snr) {
      formData.append('min_snr', String(options.min_snr));
    }
    if (options?.max_peaks_per_compound) {
      formData.append('max_peaks_per_compound', String(options.max_peaks_per_compound));
    }
    const { data } = await apiClient.post<ExtractPeaksResponse>(
      '/method-import/extract-peaks',
      formData,
      { headers: { 'Content-Type': undefined } },
    );
    return data;
  },

  trainFromPeaks: async (
    mzxmlFiles: File[],
    compoundIds: string[],
    options?: {
      methFile?: File;
      column_type?: string;
      model_type?: string;
      mz_tolerance_ppm?: number;
      min_snr?: number;
      existing_artifact_id?: string;
    } & MethodConditionOverrides,
  ) => {
    const formData = new FormData();
    for (const f of mzxmlFiles) {
      formData.append('mzxml_files', f);
    }
    formData.append('compound_ids', compoundIds.join(','));
    if (options?.methFile) {
      formData.append('meth_file', options.methFile);
    }
    formData.append('column_type', options?.column_type || 'C18');
    formData.append('model_type', options?.model_type || 'xgboost');
    if (options?.mz_tolerance_ppm) {
      formData.append('mz_tolerance_ppm', String(options.mz_tolerance_ppm));
    }
    if (options?.min_snr) {
      formData.append('min_snr', String(options.min_snr));
    }
    if (options?.existing_artifact_id) {
      formData.append('existing_artifact_id', options.existing_artifact_id);
    }
    // Method condition overrides
    if (options?.override_flow != null) {
      formData.append('override_flow', String(options.override_flow));
    }
    if (options?.override_temp != null) {
      formData.append('override_temp', String(options.override_temp));
    }
    if (options?.override_percent_b_start != null) {
      formData.append('override_percent_b_start', String(options.override_percent_b_start));
    }
    if (options?.override_percent_b_end != null) {
      formData.append('override_percent_b_end', String(options.override_percent_b_end));
    }
    if (options?.override_gradient_time != null) {
      formData.append('override_gradient_time', String(options.override_gradient_time));
    }
    if (options?.override_ph != null) {
      formData.append('override_ph', String(options.override_ph));
    }
    const { data } = await apiClient.post<TrainFromPeaksResponse>(
      '/method-import/train-from-peaks',
      formData,
      { headers: { 'Content-Type': undefined } },
    );
    return data;
  },

  // F11: CSV/TXT Chromatogram Import
  parseChromatogramCsv: async (file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    const { data } = await apiClient.post<{
      time_min: number[];
      intensity: number[];
      detector: string;
      wavelength_nm: number | null;
      sample_name: string;
      n_points: number;
      peaks: Array<{ rt_min: number; height: number; width_min: number; area: number; index: number }>;
    }>('/method-import/parse-chromatogram-csv', formData, {
      headers: { 'Content-Type': undefined },
    });
    return data;
  },
};

File: backend/app/api/routes/method_import.py
"""Method import routes: parse .meth files and extract peaks from mzXML.

Supports:
- Parsing .meth files with editable method conditions
- Multiple mzXML file uploads
- Peak extraction across all mzXML files
- Training new models or incrementally improving existing ones
"""
from __future__ import annotations

import uuid
from typing import Any

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.chem.meth_parser import MethParseError, ParsedMethod, parse_meth_file
from app.core.chem.mzxml_parser import (
    MzXmlParseError,
    extract_compound_peaks,
    parse_mzxml,
)
from app.core.ml.registry import get_artifact, list_artifacts
from app.core.ml.trainer import TrainingSample, train_model
from app.deps import CurrentUser, DBSession
from app.services import compound_service
from app.services.method_service import compute_method_signature
from pydantic import BaseModel

router = APIRouter(prefix="/method-import", tags=["method-import"])


class ParsedMethodOut(BaseModel):
    """Parsed chromatography method conditions."""
    instrument: str | None = None
    method_name: str | None = None
    column_temp_c: float | None = None
    flow_rate_ml_min: float | None = None
    solvent_a: str | None = None
    solvent_b: str | None = None
    method_end_time_min: float | None = None
    injection_volume_ul: float | None = None
    sampler_temp_c: float | None = None
    percent_b_start: float | None = None
    percent_b_end: float | None = None
    gradient_time_min: float | None = None
    gradient_table: list[dict[str, Any]] = []
    warnings: list[str] = []


class MzXmlSummaryOut(BaseModel):
    """Summary of an mzXML file."""
    filename: str = ""
    num_scans: int = 0
    num_ms1_scans: int = 0
    num_ms2_scans: int = 0
    rt_start_s: float | None = None
    rt_end_s: float | None = None
    polarity: str | None = None


class PeakDetectionResult(BaseModel):
    """Peak detection result for a single compound."""
    compound_id: str | None = None
    compound_name: str | None = None
    smiles: str | None = None
    target_mz: float | None = None
    mz_tolerance_ppm: float | None = None
    peaks: list[dict[str, Any]] = []
    xic_points: int = 0
    error: str | None = None


class ExtractPeaksResponse(BaseModel):
    """Response from peak extraction endpoint."""
    mzxml_summaries: list[MzXmlSummaryOut] = []
    results: list[PeakDetectionResult] = []
    method_conditions: ParsedMethodOut | None = None


class ModelSummary(BaseModel):
    """Summary of an existing model for selection."""
    id: str
    column_type: str
    model_type: str
    version: int
    n_samples: int
    trained_at: str


@router.post("/parse-meth", response_model=ParsedMethodOut)
async def parse_meth(
    current: CurrentUser,
    file: UploadFile = File(...),
) -> ParsedMethodOut:
    """Parse a Thermo Chromeleon .meth file and extract chromatography conditions."""
    content = await file.read()
    if not content:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Empty file")
    try:
        parsed = parse_meth_file(content)
    except MethParseError as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, str(exc)) from exc
    return ParsedMethodOut(**parsed.to_dict())


@router.post("/parse-mzxml", response_model=MzXmlSummaryOut)
async def parse_mzxml_route(
    current: CurrentUser,
    file: UploadFile = File(...),
) -> MzXmlSummaryOut:
    """Parse an mzXML file and return a summary (no peak extraction)."""
    content = await file.read()
    if not content:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Empty file")
    try:
        summary = parse_mzxml(content)
    except MzXmlParseError as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, str(exc)) from exc
    return MzXmlSummaryOut(
        filename=file.filename or "",
        num_scans=summary.num_scans,
        num_ms1_scans=summary.num_ms1_scans,
        num_ms2_scans=summary.num_ms2_scans,
        rt_start_s=summary.rt_start_s,
        rt_end_s=summary.rt_end_s,
        polarity=summary.polarity,
    )


@router.get("/models", response_model=list[ModelSummary])
async def list_models(
    db: DBSession,
    current: CurrentUser,
    column_type: str | None = None,
) -> list[ModelSummary]:
    """List existing trained models for incremental training selection."""
    artifacts = await list_artifacts(db, column_type=column_type, limit=100)
    return [
        ModelSummary(
            id=str(a.id),
            column_type=a.column_type,
            model_type=a.model_type,
            version=a.version,
            n_samples=a.n_samples,
            trained_at=a.trained_at.isoformat() if a.trained_at else "",
        )
        for a in artifacts
    ]


async def _fetch_compounds_async(
    db: AsyncSession, current: CurrentUser, compound_ids: str
) -> list[dict[str, Any]]:
    """Parse compound IDs and fetch from DB."""
    try:
        ids = [uuid.UUID(cid.strip()) for cid in compound_ids.split(",") if cid.strip()]
    except ValueError as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, f"Invalid compound ID: {exc}") from exc

    if not ids:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "No compound IDs provided")

    compounds: list[dict[str, Any]] = []
    for cid in ids:
        compound = await compound_service.get_compound(db, cid)
        if compound is None:
            raise HTTPException(status.HTTP_404_NOT_FOUND, f"Compound {cid} not found")
        if compound.owner_id != current.id and not compound.is_shared:
            raise HTTPException(status.HTTP_403_FORBIDDEN, f"Not allowed to access compound {cid}")
        compounds.append({
            "id": str(compound.id),
            "name": compound.name,
            "smiles": compound.smiles,
        })
    return compounds


def _build_method_conditions(
    method_conditions: ParsedMethod | None,
    # User-edited overrides
    override_flow: float | None = None,
    override_temp: float | None = None,
    override_percent_b_start: float | None = None,
    override_percent_b_end: float | None = None,
    override_gradient_time: float | None = None,
    override_ph: float | None = None,
) -> dict[str, float]:
    """Build method conditions dict from parsed values + user overrides."""
    ph = override_ph or 2.7
    percent_b_start = 5.0
    percent_b_end = 95.0
    gradient_time_min = 20.0
    flow_rate = 0.4
    temperature = 30.0

    if method_conditions:
        if method_conditions.percent_b_start is not None:
            percent_b_start = method_conditions.percent_b_start
        if method_conditions.percent_b_end is not None:
            percent_b_end = method_conditions.percent_b_end
        if method_conditions.gradient_time_min is not None:
            gradient_time_min = method_conditions.gradient_time_min
        if method_conditions.flow_rate_ml_min is not None:
            flow_rate = method_conditions.flow_rate_ml_min
        if method_conditions.column_temp_c is not None:
            temperature = method_conditions.column_temp_c

    # Apply user overrides
    if override_flow is not None:
        flow_rate = override_flow
    if override_temp is not None:
        temperature = override_temp
    if override_percent_b_start is not None:
        percent_b_start = override_percent_b_start
    if override_percent_b_end is not None:
        percent_b_end = override_percent_b_end
    if override_gradient_time is not None:
        gradient_time_min = override_gradient_time
    if override_ph is not None:
        ph = override_ph

    return {
        "ph": ph,
        "percent_b_start": percent_b_start,
        "percent_b_end": percent_b_end,
        "gradient_time_min": gradient_time_min,
        "flow_rate_ml_min": flow_rate,
        "temperature_c": temperature,
    }


@router.post("/extract-peaks", response_model=ExtractPeaksResponse)
async def extract_peaks(
    db: DBSession,
    current: CurrentUser,
    mzxml_files: list[UploadFile] = File(...),
    compound_ids: str = Form(...),
    meth_file: UploadFile | None = File(None),
    mz_tolerance_ppm: float = Form(10.0),
    min_snr: float = Form(3.0),
    max_peaks_per_compound: int = Form(3),
) -> ExtractPeaksResponse:
    """Extract peaks from one or more mzXML files for a list of compounds.

    Optionally parse a .meth file to include chromatography conditions.
    Compounds are specified by their database IDs (comma-separated UUIDs).
    When multiple mzXML files are provided, scans from all files are combined
    and the strongest peak across all files is reported per compound.
    """
    compounds = await _fetch_compounds_async(db, current, compound_ids)

    # Parse all mzXML files and combine scans
    all_scans: list = []
    summaries: list[MzXmlSummaryOut] = []

    for mzxml_file in mzxml_files:
        content = await mzxml_file.read()
        if not content:
            continue
        try:
            summary = parse_mzxml(content)
        except MzXmlParseError as exc:
            raise HTTPException(
                status.HTTP_400_BAD_REQUEST,
                f"Error parsing {mzxml_file.filename}: {exc}",
            ) from exc
        all_scans.extend(summary.scans)
        summaries.append(MzXmlSummaryOut(
            filename=mzxml_file.filename or "",
            num_scans=summary.num_scans,
            num_ms1_scans=summary.num_ms1_scans,
            num_ms2_scans=summary.num_ms2_scans,
            rt_start_s=summary.rt_start_s,
            rt_end_s=summary.rt_end_s,
            polarity=summary.polarity,
        ))

    if not all_scans:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "No scans found in mzXML file(s)")

    # Optionally parse .meth file
    method_conditions = None
    if meth_file:
        meth_content = await meth_file.read()
        if meth_content:
            try:
                parsed = parse_meth_file(meth_content)
                method_conditions = ParsedMethodOut(**parsed.to_dict())
            except MethParseError:
                pass  # Non-fatal

    # Extract peaks from combined scans
    peak_results = extract_compound_peaks(
        scans=all_scans,
        compounds=compounds,
        mz_tolerance_ppm=mz_tolerance_ppm,
        min_snr=min_snr,
        max_peaks_per_compound=max_peaks_per_compound,
    )

    return ExtractPeaksResponse(
        mzxml_summaries=summaries,
        results=[PeakDetectionResult(**r.to_dict()) for r in peak_results],
        method_conditions=method_conditions,
    )


@router.post("/train-from-peaks", response_model=dict)
async def train_from_peaks(
    db: DBSession,
    current: CurrentUser,
    mzxml_files: list[UploadFile] = File(...),
    compound_ids: str = Form(...),
    meth_file: UploadFile | None = File(None),
    column_type: str = Form("C18"),
    model_type: str = Form("xgboost"),
    mz_tolerance_ppm: float = Form(10.0),
    min_snr: float = Form(3.0),
    # User-editable method condition overrides
    override_flow: float | None = Form(None),
    override_temp: float | None = Form(None),
    override_percent_b_start: float | None = Form(None),
    override_percent_b_end: float | None = Form(None),
    override_gradient_time: float | None = Form(None),
    override_ph: float | None = Form(None),
    # Incremental training: provide existing artifact ID to merge data
    existing_artifact_id: str | None = Form(None),
) -> dict[str, Any]:
    """Full pipeline: parse .meth + mzXML(s), extract peaks, and train a model.

    Generates training samples from detected peaks and chromatography conditions,
    then trains a retention model. Supports:
    - Multiple mzXML files (scans combined, strongest peak per compound used)
    - User-edited method condition overrides
    - Incremental training: pass existing_artifact_id to merge new samples with
      the existing model's training data and retrain
    """
    compounds = await _fetch_compounds_async(db, current, compound_ids)

    # Parse all mzXML files and combine scans
    all_scans: list = []
    for mzxml_file in mzxml_files:
        content = await mzxml_file.read()
        if not content:
            continue
        try:
            summary = parse_mzxml(content)
        except MzXmlParseError as exc:
            raise HTTPException(
                status.HTTP_400_BAD_REQUEST,
                f"Error parsing {mzxml_file.filename}: {exc}",
            ) from exc
        all_scans.extend(summary.scans)

    if not all_scans:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "No scans found in mzXML file(s)")

    # Parse .meth file for method conditions
    parsed_method: ParsedMethod | None = None
    if meth_file:
        meth_content = await meth_file.read()
        if meth_content:
            try:
                parsed_method = parse_meth_file(meth_content)
            except MethParseError:
                pass

    # Build method conditions with user overrides
    conditions = _build_method_conditions(
        parsed_method,
        override_flow=override_flow,
        override_temp=override_temp,
        override_percent_b_start=override_percent_b_start,
        override_percent_b_end=override_percent_b_end,
        override_gradient_time=override_gradient_time,
        override_ph=override_ph,
    )

    # Extract peaks (use strongest peak per compound)
    peak_results = extract_compound_peaks(
        scans=all_scans,
        compounds=compounds,
        mz_tolerance_ppm=mz_tolerance_ppm,
        min_snr=min_snr,
        max_peaks_per_compound=1,
    )

    # Build new training samples
    new_samples: list[TrainingSample] = []
    for result in peak_results:
        if result.error or not result.peaks:
            continue
        if not result.smiles:
            continue

        peak = result.peaks[0]
        new_samples.append(TrainingSample(
            smiles=result.smiles,
            column_type=column_type,
            ph=conditions["ph"],
            percent_b_start=conditions["percent_b_start"],
            percent_b_end=conditions["percent_b_end"],
            gradient_time_min=conditions["gradient_time_min"],
            flow_rate_ml_min=conditions["flow_rate_ml_min"],
            temperature_c=conditions["temperature_c"],
            observed_rt_s=peak.retention_time_s,
        ))

    if not new_samples:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            "No training samples generated — no peaks detected for the given compounds",
        )

    # If incremental training, load existing model's samples and merge
    all_samples = new_samples
    existing_info: dict[str, Any] = {}
    if existing_artifact_id:
        try:
            artifact_id = uuid.UUID(existing_artifact_id)
        except ValueError as exc:
            raise HTTPException(
                status.HTTP_400_BAD_REQUEST, f"Invalid artifact ID: {exc}"
            ) from exc

        existing_artifact = await get_artifact(db, artifact_id)
        if existing_artifact is None:
            raise HTTPException(status.HTTP_404_NOT_FOUND, "Existing model not found")

        # Load stored runs for the same column type to get historical data
        from app.core.ml.trainer import load_stored_runs
        existing_samples = await load_stored_runs(db, column_type)
        if existing_samples:
            all_samples = existing_samples + new_samples
            existing_info = {
                "existing_samples_loaded": len(existing_samples),
                "existing_model_version": existing_artifact.version,
                "existing_model_type": existing_artifact.model_type,
            }

    # Train model
    try:
        artifact = await train_model(
            db=db,
            owner_id=current.id,
            column_type=column_type,
            model_type=model_type,
            samples=all_samples,
        )
    except Exception as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, f"Training failed: {exc}") from exc

    return {
        "artifact_id": str(artifact.id),
        "n_samples": len(all_samples),
        "n_new_samples": len(new_samples),
        "column_type": column_type,
        "model_type": model_type,
        "compounds_used": [r.compound_name or r.smiles for r in peak_results if r.peaks],
        "compounds_no_peaks": [r.compound_name or r.smiles for r in peak_results if not r.peaks],
        "incremental": existing_artifact_id is not None,
        **existing_info,
    }
